File: algorithm/leetcode910.smallest-range-ii.cpp
class Solution {
public:
    int smallestRangeII(vector<int>& A, int K) {
    	sort(A.begin(), A.end());
    	int i, n = A.size(), up, down;
    	int mindis = INT_MAX;
    	for(i = 0; i < n-1; i++)
    	{
    		up = INT_MIN, down = INT_MAX;
    		up = max(A[i]+K, A[n-1]-K);
    		down = min(A[0]+K, A[i+1]-K);
    		mindis = min(mindis, up-down);
    	}
    	return min(mindis, A[n-1]-A[0]);
    }
};

File: algorithm/Python3_version/leetcode910.smallest-range-ii.py
class Solution:
    def smallestRangeII(self, A: List[int], K: int) -> int:
        A.sort()
        mindis = float("inf")
        n = len(A)
        for i in range(n-1):
            up = max(A[i]+K, A[n-1]-K)
            down = min(A[0]+K, A[i+1]-K)
            mindis = min(mindis, up-down)
        return min(mindis, A[n-1]-A[0])# 第二种情况为同侧偏置